### PR TITLE
Add docs infrastructure

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
This PR adds the docs infrastructure for the API reference and tutorials. The templates are taken from https://github.com/Qiskit/qiskit-terra/pull/10455. 

### Details and comments
- The tutorial added is only to test that the infrastructure works. A follow-up PR will handle the actual tutorial migration from qiskit-tutorials.
- Missing:
- [x] Github Actions setup

### CODE REVIEW

File: docs/conf.py
Comment (ElePT) [inline docs/conf.py:25]: I had to add this fix for sphinx to find the path to the modules. Without it, I got `attempted relative import beyond top-level package`. If there is anything else that can be done to fix this, let me know.

**Comment by ElePT** [comment]
When I build the docs locally, the API Reference tab does not appear in the sidebar, I can only access it if I click on the `.html` directly...

Screenshot 2023-07-24 at 11 28 53

**Comment by ElePT** [comment]
I see that the CI fails on `sphinx-build -M html docs docs/_build -W -T --keep-going ` [here](https://github.com/qiskit-community/qiskit-algorithms/actions/runs/5644031946/job/15287037136?pr=13#step:7:85), but I am not able to reproduce this failure locally. Is there anything I am missing regarding tags and releasenotes?

**Comment by woodsp-ibm** [comment]
> but it looks like I lack permissions to run ignore_untagged_notes.sh].

Does it have the executable attribute set?

**Comment by woodsp-ibm** [comment]
> I am missing regarding tags and/or releasenotes

Maybe try adding a release note - could do a prelude one and put text that is similar explanation to what Terra is saying about the deprecation/move. We need something like that here anyway. Not sure if it will fix things or not.

**Comment by woodsp-ibm** [comment]
> When I build the docs locally, the API Reference tab does not appear in the sidebar, I can only access it if I click on the .html directly... Any thoughts?

What version of the qiskit-sphinx-theme do you have. I think the apps are set with this `qiskit_sphinx_theme~=1.12.0` - not sure about other settings in the conf.py without checking through them. The requirements allow > 1.11 and I know Eric has been altering to be more restrictive. Here its pulling in the latest 1.13.1 which may require some conf changes. Its based on Furo now I believe.

File: .github/workflows/deploy-docs.yml
Comment (Eric-Arellano) [inline .github/workflows/deploy-docs.yml:24]: why this part about the maintainer?

File: docs/conf.py
Comment (Eric-Arellano) [inline docs/conf.py:78]: Are you using these all? You can search for the strings like `rustworkx` to see if it's referenced elsewhere.

File: docs/conf.py
Comment (Eric-Arellano) [inline docs/conf.py:90]:
```suggestion
html_theme = "qiskit-ecosystem"
html_last_updated_fmt = "%Y/%m/%d"
html_title = f"{project} {release}"
```

This will use the much nicer Qiskit Ecosystem theme. `html_title` will make the project name in the left sidebar less noisy.

File: docs/tutorials/01_algorithms_introduction.ipynb
Comment (Eric-Arellano) [inline docs/tutorials/01_algorithms_introduction.ipynb:1]: Did this tutorial already exist? Or it's new content?

File: requirements-dev.txt
Comment (Eric-Arellano) [inline requirements-dev.txt:13]:
```suggestion
qiskit-sphinx-theme~=1.14.0rc1
```

File: tox.ini
Comment (Eric-Arellano) [inline tox.ini:60]: Are you using translations for this new repo? If so, I recommend that being a dedicated follow up PR. So you should delete this for now. 

When deleting it, also delete line 48 of setting `envdir` for `docs`.

File: .github/workflows/deploy-docs.yml
Comment (woodsp-ibm) [reply .github/workflows/deploy-docs.yml:24]: Docs deploy is manually run which publishes them to the live site - we didn't want just anyone to be able to run this.

File: .github/workflows/deploy-docs.yml
Comment (woodsp-ibm) [reply .github/workflows/deploy-docs.yml:54]: Since the VM is created for each job and requirements etv installed into that base it was all donevia makes rather than have yet another virtual env on top of that. I am trying to recall why pip install jupyter - I think it was just a workaround as some transitive dependent that used to be installed  did not. Either way this all needs to get setup in an expedited fashion and following what we have working in the apps was the chosen path. Things can be toxified if wanted later, we need to get this all setup and running asap to have the release in the next couple of days.

File: tox.ini
Comment (woodsp-ibm) [reply tox.ini:60]: I have no idea around translations - same for the apps now they are in community. Is the support wanted to be continued for that or not, if it is maybe adding it here seems reasonable too.

File: requirements-dev.txt
Comment (woodsp-ibm) [reply requirements-dev.txt:18]: Hopefully not!

File: tox.ini
Comment (Eric-Arellano) [reply tox.ini:60]: > if it is maybe adding it here seems reasonable too.

This PR already has a lot of moving parts. So to help us increase the odds of getting everything right and giving good code review, I'd personally advocate for a follow up PR to deal with translations. Translations aren't essential for the general docs infra

File: .github/workflows/deploy-docs.yml
Comment (woodsp-ibm) [reply .github/workflows/deploy-docs.yml:24]: As far as I am aware anyone can run actions - we set it up for the apps and this is what they have. One could do some more checking again. For now I want to keep it.

File: tox.ini
Comment (woodsp-ibm) [reply tox.ini:60]: I was making a more general comment about translations as I am unsure about where things will stand with that - I am fine to remove this presently - Elena had been adding just what was needed. But since it will never be used it was to me a no-op at present so I was less than bothered by it being here - but removing it is fine - I was more concerned to getting the build failures sorted than finer details like this

File: docs/conf.py
Comment (ElePT) [reply docs/conf.py:108]: Not currently, it might be used later but I'm fine removing it now and re-adding it when necessary.

File: docs/tutorials/01_algorithms_introduction.ipynb
Comment (ElePT) [reply docs/tutorials/01_algorithms_introduction.ipynb:1]: this is from qiskit-tutorials, not a history-preserving copy, but I thought we could do that in a follow-up.

File: tools/ignore_untagged_notes.sh
Comment (Eric-Arellano) [inline tools/ignore_untagged_notes.sh:3]: Bad change.

File: requirements-dev.txt
Comment (Eric-Arellano) [inline requirements-dev.txt:9]: Ditto - are these necessary?

File: requirements-dev.txt
Comment (Eric-Arellano) [inline requirements-dev.txt:14]: Necessary?

File: docs/conf.py
Comment (Eric-Arellano) [inline docs/conf.py:1]:
```suggestion
# This code is part of a Qiskit project.
```

File: docs/conf.py
Comment (Eric-Arellano) [inline docs/conf.py:3]:
```suggestion
# (C) Copyright IBM 2023.
```

File: .github/workflows/deploy-docs.yml
Comment (Eric-Arellano) [inline .github/workflows/deploy-docs.yml:3]:
```suggestion
# (C) Copyright IBM 2023.
```

File: tools/deploy_documentation.sh
Comment (Eric-Arellano) [inline tools/deploy_documentation.sh:5]:
```suggestion
# This code is part of a Qiskit project.
#
# (C) Copyright IBM 2023.
```

File: docs/index.rst
Comment (woodsp-ibm) [inline docs/index.rst:5]: I guess this really ought to have something in it as otherwise one is looking more or less a blank page when things start out. An overview for example like we have in the apps.

We should also have a Getting Started too.

(Once CI is running ok with what is already here of course!)

File: .github/workflows/deploy-docs.yml
Comment (ElePT) [reply .github/workflows/deploy-docs.yml:1]: Thanks for catching these! `make html` did not, it should probably be updated.

File: tools/ignore_untagged_notes.sh
Comment (ElePT) [inline tools/ignore_untagged_notes.sh:3]:
```suggestion
# This code is part of a Qiskit project.
```

File: requirements-dev.txt
Comment (ElePT) [reply requirements-dev.txt:9]: Nope

File: requirements-dev.txt
Comment (ElePT) [reply requirements-dev.txt:14]: probably not, I'll remove and see

Comment (ElePT) [reply requirements-dev.txt:13]: It was necessary :')

**Comment by woodsp-ibm** [comment]
@Eric-Arellano With this theme the links in the sidebar, when they refer to modules, are not matching whats in the page. In the page, since they are sub-folders of the current content, they have just the name. In the sidebar they are fully qualified paths which as you go down the structure gets longer and longer. Any thought

image

Also, the notebook that is here, as a placeholder to have all the algorithms tutorials, looking at the tutorials in the html output it does not come out properly as a tile (you can download a build artifact and see - either say the documentation zip or a tutorials one where they are built out in the content, rather than being included as-is). Are we missing some config for this theme - I did not double check myself yet - had been more concerned with getting things working.

**Comment by ElePT** [comment]
https://github.com/qiskit-community/qiskit-algorithms/pull/13#discussion_r1272272796 to not leave all of the tox comments hanging: I think it could be good to simplify the setup, but given that we now have a runnable CI and limited time before the release, I would not like to break it again just now. I would leave these improvements for a follow-up.

File: docs/conf.py
Comment (Eric-Arellano) [inline docs/conf.py:28]: Hm, should this be only 2023 for now?

```suggestion
project_copyright = f"{datetime.date.today().year}, Qiskit Algorithms Development Team"
```

File: docs/getting_started.rst
Comment (Eric-Arellano) [inline docs/getting_started.rst:110]: This won't work properly anymore. Use `.. qiskit-call-to-action-grid` for the outer part, like this:

https://github.com/Qiskit/qiskit-metapackage/blob/bb1c604a380ca2c9495121994b7d24f92aa3bfaa/docs/index.rst?plain=1#L15-L21

File: docs/index.rst
Comment (Eric-Arellano) [inline docs/index.rst:25]:
```suggestion
  About Qiskit Algorithms <self>
  Getting Started <getting_started>
  API References <apidocs/algorithms>
```

**Comment by Eric-Arellano** [comment]
> Also, the notebook that is here, as a placeholder to have all the algorithms tutorials, looking at the tutorials in the html output it does not come out properly as a tile (you can download a build artifact and see - either say the documentation zip or a tutorials one where they are built out in the content, rather than being included as-is). Are we missing some config for this theme - I did not double check myself yet - had been more concerned with getting things working.

Oof, it looks like you're hitting https://github.com/Qiskit/qiskit_sphinx_theme/issues/518. I've only been able to figure that out for one project, which was removing `sphinx-tabs`. But you're not using sphinx-tabs. It will require more debugging.

Note that it is a problem on the Pytorch theme too, only not as obvious because of some styling Pytorch uses. Experiments didn't have time to debug so copied in the CSS file directly to work around it: https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1225.

> With this theme the links in the sidebar, when they refer to modules, are not matching whats in the page. In the page, since they are sub-folders of the current content, they have just the name. In the sidebar they are fully qualified paths which as you go down the structure gets longer and longer. Any thought

I'm not reproducing that. Perhaps you have to scroll down on the left sidebar more?

Screenshot 2023-07-25 at 1 18 07 PM

**Comment by woodsp-ibm** [comment]
>>  With this theme the links in the sidebar, when they refer to modules, are not matching whats in the page. In the page, since they are sub-folders of the current content, they have just the name. In the sidebar they are fully qualified paths which as you go down the structure gets longer and longer. Any thought

> I'm not reproducing that. Perhaps you have to scroll down on the left sidebar more?

I mean the names in the sidebar eg are `qiskit_algorithms.gradients` whereas in the page for qiskit_algorithms its just `gradients` as that is the sub-module from that context. While technically its right in the sidebar its seems odd that the for the classes its just their name - ie not prefixed with qiskit_algorithms etc - yet the module is a fully qualified name

image

My take on time_evolvers would have just been to have a link to time_evolvers where variational and trotterization were modules documented there rather than hoist it to the level it is. Either way again its a fully qualified name in the sidebar rather than relative.

**Comment by Eric-Arellano** [comment]
The Furo theme highlights whatever the current page is. If you are at apidocs/algorithms.html, and scroll down your page to see the Gradients subsection, you'll have your screenshot:

Screenshot 2023-07-25 at 1 44 04 PM

That's as expected. Now, click on the Gradients link either from the page or the left sidebar, and you go to the page `stubs/qiskit_algorithms.gradients.html`, with this view:

Screenshot 2023-07-25 at 1 45 05 PM

This is all expected. It's how the base Furo theme works and e.g. pip docs.

**Comment by woodsp-ibm** [comment]
Yes but I am talking about whats in the sidebar - it just seems off that the navigation in the page has `gradients` as the entry, since its in the context of the qiskit_algorithms page,  but the corresponding nested entry in sidebar is fully qualified `qiskit_algorithm.gradients` - since its all nested that full qualification seems unneeded - the classes there as I mentioned do not have the fully qualified path name. But I get once you land on the page it links to, then then full context is there in the page.

File: docs/conf.py
Comment (woodsp-ibm) [reply docs/conf.py:28]: This code/docs is all copyright back to then so 2017-2023 is correct.

**Comment by Eric-Arellano** [comment]
> Yes but I am talking about whats in the sidebar - it just seems off that the navigation in the page has gradients as the entry, since its in the context of the qiskit_algorithms page, but the corresponding nested entry in sidebar is fully qualified qiskit_algorithm.gradients - since its all nested that full qualification seems unneeded - the classes there as I mentioned do not have the fully qualified path name. But I get once you land on the page it links to, then then full context is there in the page.

If you wanted to change the left sidebar, you would need to figure out how to get autodoc/autosummary to not set the page title for each module page to be the fully qualified module. The Sphinx theme simply uses whatever the page title is for the left sidebar.

**Comment by woodsp-ibm** [comment]
Ok so its more like alter the title of the target page to be that same name, as long as the full context is there neat the top it does not have to be in the title - none of the classes pages are that way. Anyway it was just something that jumped out at me - something for later though if we want to change things

File: docs/index.rst
Comment (Eric-Arellano) [reply docs/index.rst:25]: I don't have a strong preference :) Others use `Documentation Home` for that reason. `Overview` works for me too.

**Comment by woodsp-ibm** [comment]
@Eric-Arellano I see the search output is a little different. Before if I seached for VQE (like now in the terra docs) it gives hits like PVQD and Minimizer which it still does. The difference was before is that additional text was present so you could see the hit in context e.g search in current Qiskit (terra API ref for VQE) which may be enough for you to want to check it out or skip/

image

here all I see is and I guess I just have to assume it found VQE in that link somewhere.

image
 
 Is this output as expected, do we need to configure anything further....

**Comment by ElePT** [comment]
> Note that it is a problem on the Pytorch theme too, only not as obvious because of some styling Pytorch uses. Experiments didn't have time to debug so copied in the CSS file directly to work around it: https://github.com/Qiskit-Extensions/qiskit-experiments/pull/1225.

Thanks @Eric-Arellano. I tried copying over the CSS file and running `make html` locally and it doesn't seem to fix the issue in this case. It might be because the sources of the issue aren't the `sphinx-tabs`. As this tutorial is just a placeholder, and seeing that this might be difficult to debug, I would not worry much about it now, and maybe try again to fix it (or work around it) when the tutorials are officially migrated to this repo.

**Review by Eric-Arellano — APPROVED** [review]

**Review by woodsp-ibm — APPROVED** [review]

**Pull request merged**
